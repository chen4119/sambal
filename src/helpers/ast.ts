import {
    factory,
    createPrinter,
    createSourceFile,
    SyntaxKind,
    ScriptKind,
    ScriptTarget,
    ListFormat,
    NewLineKind,
    Expression,
    ObjectLiteralExpression,
    ArrayLiteralExpression,
    PropertyAssignment,
    NewExpression,
    Identifier,
    StringLiteral,
    NumericLiteral,
    BooleanLiteral,
    Modifier,
    VariableStatement,
    NodeFlags,
    TypeReferenceNode,
    TypeNode,
    KeywordTypeNode,
    ArrayTypeNode
} from "typescript";
import fs from "fs";

export const EXPORT_MODIFIER = factory.createModifier(SyntaxKind.ExportKeyword);

export function makeValueLiteral(value: any): Expression {
    if (Array.isArray(value)) {
        return arrayToArrayLiteral(value);
    }
    switch (typeof(value)) {
        case "string":
            return makeStringLiteral(value);
        case "number":
            return makeNumericLiteral(value);
        case "boolean":
            return makeBooleanLiteral(value);
        case "object":
            return objectToObjectLiteral(value);
        case "undefined":
            return makeIdentifier("undefined");    
        default:
            throw new Error(`Unrecognized type: ${typeof(value)}`);
    }
}

export function objectToObjectLiteral(object: any): ObjectLiteralExpression {
    const properties = [];
    for (const prop of Object.keys(object)) {
        properties.push(makePropertyAssignment(prop, makeValueLiteral(object[prop])));
    }
    return makeObjectLiteral(properties);
}

export function arrayToArrayLiteral(array: any[]): ArrayLiteralExpression {
    const values = array.map((v) => makeValueLiteral(v));
    return makeArrayLiteral(values);
}

export function makeObjectLiteral(properties): ObjectLiteralExpression {
    return factory.createObjectLiteralExpression(properties, true);
}

export function makeArrayLiteral(values: Expression[]): ArrayLiteralExpression {
    return factory.createArrayLiteralExpression(values, true);
}

export function makePropertyAssignment(propName: string, initializer): PropertyAssignment{
    return factory.createPropertyAssignment(
        makeIdentifier(propName),
        initializer
    );
}

export function makeIdentifier(text: string): Identifier {
    return factory.createIdentifier(text);
}

export function makeStringLiteral(text: string): StringLiteral {
    return factory.createStringLiteral(text);
}

export function makeNumericLiteral(number: number): NumericLiteral {
    return factory.createNumericLiteral(String(number));
}

export function makeBooleanLiteral(bool: boolean): BooleanLiteral {
    return bool ? factory.createTrue() : factory.createFalse();
}

export function makeTypeReferenceNode(typeName: string, args: TypeNode[]): TypeReferenceNode {
    return factory.createTypeReferenceNode(
        makeIdentifier(typeName),
        args
    );
}

export function makeStringKeywordTypeNode(): KeywordTypeNode<SyntaxKind.StringKeyword> {
    return factory.createKeywordTypeNode(SyntaxKind.StringKeyword);
}

export function makeUnknownKeywordTypeNode(): KeywordTypeNode<SyntaxKind.UnknownKeyword> {
    return factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword);
}

export function makeArrayTypeNode(node: KeywordTypeNode | ArrayTypeNode): ArrayTypeNode {
    return factory.createArrayTypeNode(node);
}

export function makeVariableStatement(
        modifiers: Modifier[],
        variableName: string,
        variableInitializer: Expression,
        flag = NodeFlags.Const
    ): VariableStatement {
    return factory.createVariableStatement(
        modifiers,
        factory.createVariableDeclarationList(
            [factory.createVariableDeclaration(
                makeIdentifier(variableName),
                undefined,
                undefined,
                variableInitializer
            )], flag
        )
    );
}

export function makeCallExpression(functionName: string, args: Expression[]) {
    return factory.createCallExpression(
        makeIdentifier(functionName),
        undefined,
        args
    );
}

export function makeExpressionStatement(expr: Expression) {
    return factory.createExpressionStatement(expr);
}

export function makeEqualsBinaryExpression(leftExpr: Expression, rightExpr: Expression) {
    return factory.createBinaryExpression(
        leftExpr,
        factory.createToken(SyntaxKind.EqualsToken),
        rightExpr
    );
}

export function makePropertyAccess(expr: Expression, propName: string) {
    return factory.createPropertyAccessExpression(expr, makeIdentifier(propName));
}

export function makeEnum(modifiers: Modifier[], enumName: string, values: string[]) {
    return factory.createEnumDeclaration(
        undefined,
        modifiers,
        makeIdentifier(enumName),
        values.map((v) => factory.createEnumMember(makeIdentifier(v), undefined))
    );
}

export function makeNew(expression: Expression, args: Expression[]): NewExpression {
    return factory.createNewExpression(
        expression,
        undefined,
        args
    );
}

export function writeJavascript(statements, output) {
    const tsPrinter = createPrinter({
        newLine: NewLineKind.LineFeed
    });
    const sourceFile = createSourceFile(
        output,
        "",
        ScriptTarget.ES2015,
        false,
        ScriptKind.JS
    );
    const tsSource = tsPrinter.printList(
        ListFormat.MultiLine,
        factory.createNodeArray(statements),
        sourceFile
    );

    fs.writeFileSync(output, tsSource, "utf-8");
}