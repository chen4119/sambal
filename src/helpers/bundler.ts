import webpack, { Configuration } from "webpack";
import path from "path";
import nodeExternals from "webpack-node-externals";
import {
    CACHE_FOLDER,
    OnBundleChanged
} from "./constant";
import { log } from "./log";

const outputConfig = {
    library: {
        type: "umd"
    },
    globalObject: 'this'
};

const nodeJsConfig = {
    externalsPresets: { node: true },
    externals: [nodeExternals()]
};

const webpackConfig = {
    resolve: {
        extensions: [".js", ".jsx"]
    },
    module: {
        rules: [
            {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env", "@babel/preset-react"],
                        plugins: ["@babel/plugin-transform-runtime"]
                    }
                }
            },
            {
                test: /\.css$/,
                use: ['style-loader', 'css-loader']
            },
            {
                test: /\.(png|svg|jpg|jpeg|gif)$/i,
                type: 'asset/resource',
            },
            {
                // Apply rule for fonts files
                test: /\.(svg|woff|woff2|ttf|otf|eot)$/,
                type: "asset/inline"
            },
        ]
    }
};

export async function bundleSambalFile(entry: string, outputPath: string, isExcludeNode: boolean = true) {
    return new Promise<void>((resolve, reject) => {
        const config: Configuration = {
            ...webpackConfig,
            ...isExcludeNode ? nodeJsConfig : { externalsPresets: { node: true } },
            entry: entry,
            mode: "production",
            output: {
                ...outputConfig,
                filename: path.basename(entry),
                path: outputPath
            }
        };
        const compiler = webpack(config);
        compiler.run((err, stats) => {
            parseStats(err, stats, () => {
                if (!err) {
                    resolve();
                } else {
                    reject(err);
                }
            });
        });
    });
}

export function watchSambalFile(entry: string, onChange: OnBundleChanged) {
    const config: Configuration = {
        ...webpackConfig,
        ...nodeJsConfig,
        entry: entry,
        mode: "development",
        devtool: "eval-source-map",
        output: {
            ...outputConfig,
            filename: "[name].[contenthash].js",
            path: path.resolve(process.cwd(), `./${CACHE_FOLDER}/watch`)
        }
    };
    const compiler = webpack(config);

    return compiler.watch({
        aggregateTimeout: 300,
        poll: 1000
    }, (err, stats) => {
        parseStats(err, stats, onChange);
    });
}

export async function bundleBrowserPackage(bundleConfig: any, outputPath: string) {
    return new Promise<object>((resolve, reject) => {
        const config: Configuration = {
            ...webpackConfig,
            entry: bundleConfig.entry,
            mode: "production",
            output: {
                ...outputConfig,
                filename: "[name].[contenthash].js",
                path: outputPath
            }
        };
        const compiler = webpack(config);
        compiler.run((err, stats) => {
            parseStats(err, stats, (isError, entry) => {
                if (!err) {
                    resolve(entry);
                } else {
                    reject(err);
                }
            });
        });
    });
}

export function getDevServerBrowserCompiler(bundleConfig: any, listener) {
    const config: Configuration = {
        ...webpackConfig,
        entry: bundleConfig.entry,
        mode: "development",
        devtool: "eval-source-map",
        output: {
            ...outputConfig,
            filename: "[name].[contenthash].js"
        },
        plugins: [listener]
    };
    return webpack(config);
}

export function parseBundleFilename(stats) {
    const entries: any = {};
    for (const key of Object.keys(stats.entrypoints)) {
        const mainAsset = stats.entrypoints[key].assets[0];
        entries[key] = mainAsset.name;
    }
    return entries;
}

function parseStats(err, stats, onChange: OnBundleChanged) {
    const info = stats.toJson();
    const entry = parseBundleFilename(info);
    if (err) {
        log.error(err.stack || err.message);
        onChange(true, entry);
        return;
    }
    if (stats.hasErrors()) {
        // log.error(info.errors);
        for (const err of info.errors) {
            log.error(err.message);
        }
        onChange(true, entry);
        return;
    }
    if (stats.hasWarnings()) {
        // console.warn("%O", ...info.warnings);
        for (const warning of info.warnings) {
            log.warn(warning.message);
        }
    }
    onChange(false, entry);
}