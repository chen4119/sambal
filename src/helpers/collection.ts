import { JSONLD_ID } from "sambal-jsonld";
import {
    NULL_VALUE,
    EMPTY_VALUE
} from "./constant";
import {
    isJsDate
} from "./util";


// const MAX_INDEX_VALUE_CHAR = 600;

/*
export function mapJsonToPartitions(expandedJsonLd, groupBy) {
    const groupByValues = [];
    for (const groupByField of groupBy) {
        groupByValues.push(expandedJsonLd[groupByField]);
    }
    return getPartitionKeys(groupBy, groupByValues);
}

function getPartitionKeys(groupByKeys, groupByValues) {
    const partitionKeys = [];
    const indexes = groupByValues.map(value => Array.isArray(value) ? value.length : 0);
    let isDone = false;
    while (!isDone) {
        const partitionKey = [];
        for (let i = 0; i < groupByKeys.length; i++) {
            const value = groupByValues[i];
            if (Array.isArray(value)) {
                partitionKey.push(normalizeValue(value[--indexes[i]]));
            } else {
                partitionKey.push(normalizeValue(value));
            }
        }
        if (partitionKey.join("").length > MAX_INDEX_VALUE_CHAR) {
            throw new Error(`Partition key cannot excceed ${MAX_INDEX_VALUE_CHAR} chars`);
        }
        partitionKeys.push(partitionKey);
        isDone = indexes.filter(i => i > 0).length === 0; 
    }
    return partitionKeys;
}*/

function normalizeValue(value) {
    if (value === NULL_VALUE || value === EMPTY_VALUE) {
        throw new Error(`${value} is reserved`);
    } else if (typeof(value) === "undefined" || value === null) {
        return NULL_VALUE;
    } else if (typeof(value) === "string" && value.length === 0) {
        return EMPTY_VALUE;
    } else if (isJsDate(value)) {
        return value.getTime();
    } else if (typeof(value) === "object") {
        if (!value[JSONLD_ID]) {
            throw new Error("No complex index value allowed");
        }
        return value[JSONLD_ID];
    }
    return value;
}

export function getSortKey(jsonLd, propName: string) {
    return normalizeValue(jsonLd[propName]);
}

