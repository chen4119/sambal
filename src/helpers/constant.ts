
/*
export const WebElement = {
    SiteNavigation: "SiteNavigation",
    AdBlock: "AdBlock",
    Footer: "Footer",
    Header: "Header",
    SideBar: "SideBar",
    // Table: "Table"
}

export const Action = {
    Search: "Search",
    Review: "Review",
    Like: "Like",
    Listen: "Listen",
    View: "View",
    Watch: "Watch",
    Subscribe: "Subscribe",
    Comment: "Comment",
    Reply: "Reply",
    Share: "Share",
}*/

export const SORT_ASC = "asc";
export const SORT_DESC = "desc";

export type SortBy = {
    prop: string,
    order: "asc" | "desc"
}

export type Theme = {
    name: string,
    options: object
};

export type Collection = {
    "@id": string,
    "@type"?: string,
    src: string | string[],
    groupBy?: string[],
    sortBy?: SortBy
}

export type PageNode = {
    "@id": string,
    "@type": string,
    url: string,
    mainEntity: any,
    hasPart: any[],
    [key: string]: any
}

export type OnBundleChanged = (isError: boolean, entry: any) => void;

export interface IHtmlSerializer {
    toHtml(renderResult: unknown);
}

export const CACHE_FOLDER = ".sambal";
export const OUTPUT_FOLDER = "public";
export const SAMBAL_ENTRY_FILE = "sambal.entry.js";
export const SAMBAL_SITE_FILE = "sambal.site.js";
export const WEBSOCKET_ADDR = "ws://localhost:3001/";

// collection
export const NULL_VALUE = "__null__";
export const EMPTY_VALUE = "__empty__";
