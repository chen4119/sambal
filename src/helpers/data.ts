import axios, { AxiosResponse } from "axios";
import yaml from "js-yaml";
// import marked from "marked";
import glob from "glob";
import {
    getAbsFilePath,
    isValidRelativePath,
    readFileAsBuffer,
    readTextFile,
    frontMatter,
    safeParseJson,
    getFileExt
} from "./util";
import { isAbsUri } from "sambal-jsonld";

enum SUPPORTED_CONTENT_TYPE {
    yaml,
    markdown,
    json,
    image
}

export function searchFiles(query: string | string[]): string[] {
    if (Array.isArray(query)) {
        const matchSet = query.reduce((accumulator, value) => {
            searchFiles(value).forEach(m => accumulator.add(m));
            return accumulator;
        }, new Set<string>());
        return Array.from(matchSet);
    }
    if (isAbsUri(query)) {
        return [query];
    }
    const matches = glob.sync(query);
    return matches.filter(m => isValidRelativePath(m));
}

export function inferUrl(uri: string) {
    if (!uri || typeof(uri) !== "string") {
        return "";
    }
    if (uri.startsWith("/pages")) {
        let url = uri.substring(6);
        if (isSupportedFile(url)) {
            url = url.substring(0, url.lastIndexOf("."));
        }
        if (url === "/index") {
            return "/";
        }
        if (url.endsWith("/index")) {
            url = url.substring(0, url.length - 6);
        }
        return url;
    }
    return uri;
}

export async function loadLocalFile(src: string) {
    if (!isSupportedFile(src)) {
        throw new Error(`Unsupported file type: ${src}`);
    }

    const contentType = getLocalFileContentType(src);
    let content;
    if (contentType === SUPPORTED_CONTENT_TYPE.image) {
        content = await readFileAsBuffer(getAbsFilePath(src));
    } else {
        content = await readTextFile(getAbsFilePath(src));
    }
    return parseContent(content, contentType);
}

export async function loadRemoteFile(src: string) {
    const response = await axios.get(src);
    return parseContent(response.data, getAxiosResponseContentType(response));
}

const IMAGE_EXT_REGEX = /.+(.jpg|.jpeg|.gif|.png|.webp)$/i;
const SUPPORTED_FILE_EXT_REGEX = /.+(.yaml|.yml|.json|.md)$/i;

export function isImageFile(src: string) {
    if (!src) {
        return false;
    }
    return src.match(IMAGE_EXT_REGEX);
}

export function isSupportedFile(src: string) {
    if (!src) {
        return false;
    }
    return isAbsUri(src) || 
    isImageFile(src) ||
    src.match(SUPPORTED_FILE_EXT_REGEX);
}

function getAxiosResponseContentType(response: AxiosResponse<any>): SUPPORTED_CONTENT_TYPE {
    const contentType = response.headers["content-type"];
    if (contentType) {
        switch (contentType) {
            case "text/yaml":
            case "text/yml":
            case "application/yaml":
            case "application/yml":
                return SUPPORTED_CONTENT_TYPE.yaml;
            case "application/json":
            case "application/ld+json":
                return SUPPORTED_CONTENT_TYPE.json;
            case "text/markdown":
                return SUPPORTED_CONTENT_TYPE.markdown;
            case "image/jpeg":
            case "image/webp":
            case "image/gif":
            case "image/png":
                return SUPPORTED_CONTENT_TYPE.image;
        }
    }
    throw `Unsupported content type ${contentType}`;
}

function getLocalFileContentType(src: string): SUPPORTED_CONTENT_TYPE {
    const ext = getFileExt(src);
    switch (ext) {
        case "yaml":
        case "yml":
            return SUPPORTED_CONTENT_TYPE.yaml;
        case "json":
            return SUPPORTED_CONTENT_TYPE.json;
        case "md":
            return SUPPORTED_CONTENT_TYPE.markdown;
        case "jpg":
        case "jpeg":
        case "webp":
        case "gif":
        case "png":
            return SUPPORTED_CONTENT_TYPE.image;
        default:
            throw `Unsupported file type with extension ${ext}`;
    }
}

// TODO: ensure markdown is only for creative work!
function parseContent(content: any, contentType: SUPPORTED_CONTENT_TYPE) {
    try {
        let md;
        switch (contentType) {
            case SUPPORTED_CONTENT_TYPE.yaml:
                return yaml.load(content);
            case SUPPORTED_CONTENT_TYPE.json:
                return typeof(content) === "string" ? safeParseJson(content) : content;
            case SUPPORTED_CONTENT_TYPE.markdown:
                md = frontMatter(content);
                return {
                    ...md.frontMatter ? yaml.load(md.frontMatter) as object : {},
                    text: md.content,
                    encodingFormat: "text/markdown"
                }
            case SUPPORTED_CONTENT_TYPE.image:
                return content;
                // return Buffer.from(content, 'binary');
            default:
                throw `Unsupported content type ${contentType}.  Expecting yaml, json or markdown`;
        }
    } catch (e) {
        throw `Unable to parse content: ${e}`;
    }
}