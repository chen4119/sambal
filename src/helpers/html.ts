import {Parser} from "htmlparser2";
import {Handler} from "htmlparser2/lib/Parser";

// TODO: spacing not used.  Remove?
const SPACING = 4;
const SAMBAL_SLOT_NAME = "sambal-slot";
const SELF_CLOSING_TAGS = [
    'area',
    'base',
    'br',
    'embed',
    'hr',
    'iframe',
    'img',
    'input',
    'link',
    'meta',
    'param',
    'source',
    'track',
    SAMBAL_SLOT_NAME
];

type HtmlAttributes = {
    [propName: string]: string
};

export async function replaceScriptSrc(srcHtml: string, replace: (src: string) => string) {
    return new Promise<string>((resolve, reject) => {
        const handler = getHtmlSerializer(resolve, reject, null, (name, attribs) => {
            if (name === "script" && attribs.src) {
                return {
                    ...attribs,
                    src: replace(attribs.src)
                };
            }
            return attribs;
        });
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

export async function hydrateSambalSlots(
    srcHtml: string,
    render: (id: string) => Promise<string>) {

    const slots = await searchForSambalSlots(srcHtml);
    for (const slot of slots) {
        slot.html = await render(slot.id);
    }

    return new Promise<string>((resolve, reject) => {
        let slotIndex = -1;
        const handler = getHtmlSerializer(resolve, reject, (name, attribs) => {
            if (name === SAMBAL_SLOT_NAME) {
                slotIndex++;
                return slots[slotIndex].html ? slots[slotIndex].html : "";
            }
            return null;
        }, null);
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

function getHtmlSerializer(
    resolve,
    reject,
    replace: (name: string, attribs: HtmlAttributes) => string,
    editAttribs: (name: string, attribs: HtmlAttributes) => HtmlAttributes) {
    const stack: string[] = [];
    let html = '';
    let deleteTag: string = null; 
    const handler: Partial<Handler> = {
        onopentag: (name,  attribs) => {
            stack.unshift(name);
            if (deleteTag != null) {
                return;
            }
            const replaceHtml = replace ? replace(name, attribs) : null;
            if (replaceHtml !== null) {
                deleteTag = name;
                html += replaceHtml;
            } else {
                const updatedAttribs = editAttribs ? editAttribs(name, attribs) : attribs;
                html += addOpeningTag(0, name, updatedAttribs);
            }
            
        },
        ontext:(data) => {
            if (deleteTag !== null) {
                return;
            }
            html += data;
        },
        onclosetag: (name) => {
            const tagName = stack.shift();
            if (deleteTag === tagName) {
                deleteTag = null;
            } else {
                html += addClosingTag(0, name);
            }
        },
        onprocessinginstruction: (name, data) => {
            html += addProcessingInstruction(0, data);
        },
        onerror: (err) => {
            reject(err);
        },
        onend: () => {
            resolve(html);
        }
    };
    return handler;
}

async function searchForSambalSlots(srcHtml: string) {
    return new Promise<any[]>((resolve, reject) => {
        const slots = [];
        const handler: Partial<Handler> = {
            onopentag: (name,  attribs) => {
                if (name === SAMBAL_SLOT_NAME) {
                    slots.push({
                        id: attribs.link
                    });
                }
            },
            onerror: (err) => {
                reject(err);
            },
            onend: () => {
                resolve(slots);
            }
        };
        const parser = new Parser(handler);
        parser.write(srcHtml);
        parser.end();
    });
}

function addOpeningTag(indent: number, name: string, attribs?: HtmlAttributes) {
    let attributes = '';
    if (attribs) {
        for (const key of Object.keys(attribs)) {
            attributes += ` ${key}="${attribs[key]}"`;
        }
    }
    return `${makeIndent(indent * SPACING)}<${name}${attributes}>`;
}

function addClosingTag(indent: number, name: string) {
    if (SELF_CLOSING_TAGS.indexOf(name) >= 0) {
        return '';
    }
    return `${makeIndent(indent * SPACING)}</${name}>`;
}

function addProcessingInstruction(indent: number, data: string) {
    return `${makeIndent(indent * SPACING)}<${data}>\n`;
}

function makeIndent(numSpaces: number) {
    let indent = '';
    for (let i = 0; i < numSpaces; i++) {
        indent += ' ';
    }
    return indent;
}
