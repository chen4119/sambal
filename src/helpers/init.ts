
export function initSambalEntry(): string {
    return`
import React from "react";

const Head = () => (
    <head>
        <meta charSet="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>
        <meta httpEquiv="X-UA-Compatible" content="ie=edge"/>
    </head>
);

export function renderPage({ page, options }) {
    const { mainEntity } = page;
    return (
        <html>
            <Head/>
            <body>
                <h1>{mainEntity.headline}</h1>
                <p>By {mainEntity.author.name}</p>
                <p>Google Analytics Id {options.googleAnalyticsId}</p>
            </body>
        </html>
    );
}

export const defaultOptions = {
    googleAnalyticsId: "UA-123"
}
`.trim();
}

export function initSambalSite(): string {
    return`
export const siteConfig = {
    baseUrl: "https://example.com",
    collections: [
        // For more info - https://www.sambal.dev/docs/core/collection/
    ],
    imageTransforms: [
        // For more info - https://www.sambal.dev/docs/core/image/
    ],
    resolvers: [
        // For more info - https://www.sambal.dev/docs/core/uri-resolver/
    ],
    // theme: "sambal-ui-material"
};
`.trim();
}

export function initPerson(): string {
    return`
"@type": Person
name: Blue Devil
email: blue@devil.com
familyName: Devil
givenName: Blue
`.trim();
}

export function initBlogpost(authorIRI: string): string {
    return`
---
"@type": BlogPosting
headline: My first blog
author:
    "@id": ${authorIRI}
dateCreated: 2002-05-01
keywords: ["tag1", "tag2"]
---

First post on Sambal!
`.trim();
}