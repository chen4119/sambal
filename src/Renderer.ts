import {
    getAbsFilePath,
    isObjectLiteral
} from "./helpers/util";
import {
    SAMBAL_ENTRY_FILE,
    Theme,
    WebPage,
    JS_FOLDER
} from "./helpers/constant";
import Bundler from "./Bundler";
import Html from "./Html";
import { log } from "./helpers/log";
import { template } from "./ui/template";

type UI = {
    renderPage: (props: {
        page: unknown,
        options?: unknown
    }) => Promise<string>,
    defaultOptions?: object
}

type Asset = {
    entry: string,
    onPages: Set<string>
}

type DevServerChangeHandler = (urls: string[]) => void;

export default class Renderer {
    private renderer: UI;
    private bundler: Bundler;
    private rootDir: string;
    private assets: Map<string, Asset>;
    // Map css source url() to dest url()
    private cssAssetMap: Map<string, string>;
    private devServerChangeHandler: DevServerChangeHandler;

    constructor(
        private baseUrl: string,
        private publicPath: string,
        private entryFile: string,
        private theme: string | Theme) {
        this.assets = new Map<string, Asset>();
        this.cssAssetMap = new Map<string, string>();
        this.rootDir = "."; // default to project root folder
    }

    async bundle() {
        let uiEntryFile = this.entryFile;
        let themeFolder;
        // Theme has precedence over entry file
        if (this.theme) {
            if (typeof(this.theme) === "string") {
                themeFolder = this.theme;
                uiEntryFile = getAbsFilePath(`${this.theme}/${SAMBAL_ENTRY_FILE}`);
            } else if (isObjectLiteral(this.theme)) {
                themeFolder = this.theme.name;
                uiEntryFile = getAbsFilePath(`${this.theme.name}/${SAMBAL_ENTRY_FILE}`);
            }
            this.rootDir = themeFolder;
        }
        log.info("Bundling sambal.entry.js...");
        const moduleEntry = await Bundler.bundleSambalFile(uiEntryFile);
        this.renderer = require(moduleEntry);
        if (!this.renderer) {
            throw new Error("No html renderer available.  Implement sambal.entry.js or specify a theme in sambal.site.js");
        }
    }

    async devInit(onChangeHandler: DevServerChangeHandler) {
        this.devServerChangeHandler = onChangeHandler;
        // bundler initialized only in dev mode
        this.bundler = new Bundler(this.onAssetChanged.bind(this));
        // Theme has precedence over entry file
        if (this.theme) {
            // bundle theme entry file, no need to watch
            await this.bundle();
        } else if (this.entryFile) {
            const moduleEntry = await this.bundler.watchSambalFile(this.entryFile);
            this.renderer = require(moduleEntry);
        }
        
        if (!this.renderer) {
            throw new Error("No html renderer available.  Implement sambal.entry.js or specify a theme in sambal.site.js");
        }
    }

    async stop() {
        this.bundler.stop();
    }

    private onAssetChanged(uri: string, entry: string) {
        if (this.assets.has(uri)) {
            this.assets.get(uri).entry = entry;
            this.devServerChangeHandler([...this.assets.get(uri).onPages.values()]);
        }
    }

    async renderPage(page: WebPage) {
        try {
            const options = {
                ...this.getDefaultOptions(this.renderer),
                ...isObjectLiteral(this.theme) ? (this.theme as Theme).options : {}
            };
            
            const html = await this.renderer.renderPage({ 
                page: page,
                options: options
            });

            if (html) {
                return await this.postProcessHtml(page, html);
            }
            return await this.renderErrorPage("Nothing rendered");
        } catch (e) {
            return await this.renderErrorPage(e);
        }
    }

    async renderErrorPage(e) {
        const renderError = (error) => {
            return template`
                <h1>${error.message}</h1>
                <p>${error.toString()}</p>
            `;
        };
        return await template`
            <html>
                <head>
                    <meta charset="UTF-8" />
                    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>
                    <meta http-equiv="X-UA-Compatible" content="ie=edge"/>
                    <base href="/">
                </head>
                <body>
                    ${Array.isArray(e) ?
                        e.map(error => renderError(error)) :
                        renderError(e)}
                </body>
            </html>
        `;
    }

    private async postProcessHtml(page: WebPage, html: string) {
        const htmlPage = new Html(html);
        for (const src of htmlPage.jsSources) {
            const entry = await this.bundleJs(src, page.url);
            htmlPage.replaceJsScriptSrc(src, this.relativeToRoot(entry));
        }
        for (const src of htmlPage.styleSheets) {
            const entry = await this.bundleCss(src, page.url);
            htmlPage.replaceStyleSheetSrc(src, this.relativeToRoot(entry));
        }
        await htmlPage.bundleStyles(this.bundleInlineStyle.bind(this));
        htmlPage.addSchemaJsonLd(page.mainEntity);
        htmlPage.addMetaTags(this.baseUrl, page);
        return htmlPage.serialize();
    }

    private relativeToRoot(path: string) {
        return path.startsWith("/") ? path : `/${path}`;
    }

    private async bundleInlineStyle(css: string) {
        return await Bundler.bundleStyle(this.cssAssetMap, css, this.publicPath);
    }

    private async bundleCss(cssPath: string, pageUrl: string) {
        const resolvedPath = this.resolvePath(cssPath);
        let entry;
        
        // If this.bundler inited, means dev mode
        if (this.bundler) {
            entry = await this.bundler.watchCssFile(this.cssAssetMap, resolvedPath, this.publicPath);
        } else {
            if (this.assets.has(resolvedPath)) {
                entry = this.assets.get(resolvedPath).entry;
            } else {
                entry = await Bundler.bundleCssFile(this.cssAssetMap, resolvedPath, this.publicPath);
                entry = entry.substring(this.publicPath.length);
            }
        }

        this.addToAssets(resolvedPath, entry, pageUrl);
        return entry;
    }

    private async bundleJs(jsPath: string, pageUrl: string) {
        const resolvedPath = this.resolvePath(jsPath);
        let entry;
        
        // If this.bundler inited, means dev mode
        if (this.bundler) {
            entry = await this.bundler.watchBrowserBundle(resolvedPath, this.getJsOutputPath());
        } else {
            if (this.assets.has(resolvedPath)) {
                entry = this.assets.get(resolvedPath).entry;
            } else {
                entry = await Bundler.bundleBrowserPackage(resolvedPath, this.getJsOutputPath());
                entry = entry.substring(this.publicPath.length);
            }
        }

        this.addToAssets(resolvedPath, entry, pageUrl);
        return entry;
    }

    private getJsOutputPath() {
        return `${this.publicPath}/${JS_FOLDER}`;
    }

    private resolvePath(filePath: string) {
        if (filePath.indexOf("node_modules/") >= 0) {
            return filePath;
        }
        return `${this.rootDir}/${filePath}`;
    }

    private addToAssets(assetUri: string, assetEntry: string, pageUrl: string) {
        if (this.assets.has(assetUri)) {
            const asset = this.assets.get(assetUri);
            asset.entry = assetEntry;
            asset.onPages.add(pageUrl);
        } else {
            this.assets.set(assetUri, {
                entry: assetEntry,
                onPages: new Set([pageUrl])
            });
        }
    }

    private getDefaultOptions(renderer: UI) {
        if (renderer.defaultOptions && isObjectLiteral(renderer.defaultOptions)) {
            return renderer.defaultOptions;
        }
        return {};
    }
}