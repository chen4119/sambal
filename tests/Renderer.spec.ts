import shelljs from "shelljs";
import Renderer from "../src/Renderer";
import { OUTPUT_FOLDER } from "../src/helpers/constant";

describe("Renderer", () => {
    const baseUrl = "https://example.com";
    let renderer: Renderer;

    beforeEach(async () => {
        renderer = new Renderer(null, "mock-theme");
    });

    afterEach(async () => {
        shelljs.rm("-rf", OUTPUT_FOLDER);
    });

    it('copy theme bundle to /public', async () => {
        await renderer.build("/js");
        expect(shelljs.test('-f', "public/js/mock-theme/client.123.js")).toBeTruthy();
    });

    it('render using mock-theme', async () => {
        await renderer.build("/js");
        const result = await renderer.renderPage({
            "@type": "WebPage",
            url: "/dummy",
            mainEntity: {
                "@type": "Person",
                name: "John Smith"
            },
        }, "/js");
        expect(result).toMatchSnapshot();
    });



});