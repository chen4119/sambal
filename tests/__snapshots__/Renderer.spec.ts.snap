// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer render using mock-theme 1`] = `
"
        <html>
            <head>
                <script src=\\"/js/main.7052bf0ac25de26ce68e.js\\"></script>
                <script src=\\"/js/main.aa4bac5091606d7c3286.js\\"></script>
                <link rel=\\"stylesheet\\" href=\\"/mock-theme/theme.Lpjvzyneslx4vFRCS_x-VHb_rf0.css\\">
                <link rel=\\"stylesheet\\" href=\\"/node_modules/@fontsource/roboto/300.fc7jZ3ovBRwuUs8Nzx6S0gvHi0k.css\\">
                <style>
                    pre[class*=&quot;language-&quot;] {
                        margin-top: 24px;
                        margin-bottom: 24px;
                    }
                </style>
            <script type=\\"application/ld+json\\">{
    \\"@context\\": \\"https://schema.org\\",
    \\"@type\\": \\"Person\\",
    \\"name\\": \\"John Smith\\"
}</script><link rel=\\"canonical\\" href=\\"https://example.com/dummy\\"><meta name=\\"og:url\\" content=\\"https://example.com/dummy\\"><title>John Smith</title><meta name=\\"twitter:title\\" content=\\"John Smith\\"><meta name=\\"og:title\\" content=\\"John Smith\\"></head>
            <body>
                <h1>Hello from mock theme</h1>
                <p>test1</p>
            </body>
        </html>
    "
`;
